resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -829,50 +967,175 @@ void add(DnsRecord r) {
+                    servers++;
+
+                    // Try to resolve via cache as we had no ADDITIONAL entry for the server.
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries != null && !entries.isEmpty()) {
+                        InetAddress address = entries.get(0).address();
+
+                        // If address is null we have a resolution failure cached so just use an unresolved address.
+                        if (address != null) {
+                            serverName.update(parent.newRedirectServerAddress(address));
+
+                            for (int i = 1; i < entries.size(); i++) {
+                                address = entries.get(i).address();
+
+                                assert address != null;
+
+                                AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                server.next = serverName.next;
+                                serverName.next = server;
+                                serverName = server;
+                                serverName.update(parent.newRedirectServerAddress(address));
+
+                                servers++;
+                            }
+                        }
+                    }
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        private static void cacheUnresolved(
+                AuthoritativeNameServer server, AuthoritativeDnsServerCache authoritativeCache, EventLoop loop) {
+            // As we have no idea about how long the TTL should be we will just cache the unresolved
+            // address
+
+            // We still want to cached the unresolved address
+            server.address = InetSocketAddress.createUnresolved(
+                    server.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+
+            // Cache the server now.
+            cache(server, authoritativeCache, loop);
+        }
+
+        private static void cache(AuthoritativeNameServer server, AuthoritativeDnsServerCache cache, EventLoop loop) {
+            // Cache NS record if not for a root server as we should never cache for root servers.
+            if (!server.isRootServer() && server.address != null) {
+                // If we resolved the AuthoritativeNameServer via the DnsCache before we should not cache the
+                // resolved address as we don't have a good idea about the TTL to apply. We will just cache
+                // the unresolved address for now. We will then query the cache again the next time we try to
+                // use it for a query which will do the correct thing in terms of respecting the TTL of the
+                // A / AAAA record.
+                InetSocketAddress addressToCache = server.ttl == Long.MIN_VALUE ?
```

Comment:
`server.ttl == Long.MIN_VALUE` -> is this still necessary? Can't we just use `server.address`?